clang/tools/clang-linker-wrapper/SYCLOffloadWrapper.cpp

```diff
@@ -0,0 +1,748 @@
+  //      PropertySetEnd0               /*EntriesEnd*/
+  ///   },
+  ///   ...
+  /// };
+  ///
+  /// static const __tgt_bin_desc FatbinDesc = {
+  ///   Version,                             /*Version*/
+  ///   sizeof(Images) / sizeof(Images[0]),  /*NumDeviceImages*/
+  ///   Images,                              /*DeviceImages*/
+  ///   __start_offloading_entries,      /*HostEntriesBegin*/
+  ///   __stop_offloading_entries        /*HostEntriesEnd*/
+  /// };
+  ///
+  /// Global variable that represents FatbinDesc is returned.
+  GlobalVariable *createFatbinDesc(SmallVector<SYCLImage> &Images) {
+    const std::string OffloadKindTag =
+        (Twine(".") + "sycl" + Twine("_offloading.")).str();
+    SmallVector<Constant *> WrappedImages;
+    WrappedImages.reserve(Images.size());
+    for (size_t i = 0; i != Images.size(); ++i) {
+      WrappedImages.push_back(wrapImage(Images[i], i, OffloadKindTag));
+      Images[i].Image.clear(); // This is just for economy of RAM.
+    }
+
+    return combineWrappedImages(WrappedImages, OffloadKindTag);
+  }
+
+  void createRegisterFatbinFunction(GlobalVariable *FatbinDesc) {
+    auto *FuncTy = FunctionType::get(Type::getVoidTy(C), /*isVarArg*/ false);
+    auto *Func = Function::Create(FuncTy, GlobalValue::InternalLinkage,
+                                  Twine("sycl") + ".descriptor_reg", &M);
+    Func->setSection(".text.startup");
+
+    // Get RegFuncName function declaration.
+    auto *RegFuncTy =
+        FunctionType::get(Type::getVoidTy(C), PointerType::getUnqual(C),
+                          /*isVarArg=*/false);
+    FunctionCallee RegFuncC =
+        M.getOrInsertFunction("__sycl_register_lib", RegFuncTy);
+
+    // Construct function body
+    IRBuilder Builder(BasicBlock::Create(C, "entry", Func));
+    Builder.CreateCall(RegFuncC, FatbinDesc);
+    Builder.CreateRetVoid();
+
+    // Add this function to constructors.
+    appendToGlobalCtors(M, Func, /*Priority*/ 1);
+  }
+
+  void createUnregisterFunction(GlobalVariable *FatbinDesc) {
+    auto *FuncTy = FunctionType::get(Type::getVoidTy(C), /*isVarArg*/ false);
+    auto *Func = Function::Create(FuncTy, GlobalValue::InternalLinkage,
+                                  "sycl.descriptor_unreg", &M);
+    Func->setSection(".text.startup");
+
+    // Get UnregFuncName function declaration.
+    auto *UnRegFuncTy =
+        FunctionType::get(Type::getVoidTy(C), PointerType::getUnqual(C),
+                          /*isVarArg=*/false);
+    FunctionCallee UnRegFuncC =
+        M.getOrInsertFunction("__sycl_unregister_lib", UnRegFuncTy);
+
+    // Construct function body
+    IRBuilder<> Builder(BasicBlock::Create(C, "entry", Func));
+    Builder.CreateCall(UnRegFuncC, FatbinDesc);
+    Builder.CreateRetVoid();
+
+    // Add this function to global destructors.
+    appendToGlobalDtors(M, Func, /*Priority*/ 1);
+  }
+}; // end of Helper
+
+} // anonymous namespace
+
+Error wrapSYCLBinaries(llvm::Module &M, SmallVector<SYCLImage> &Images,
```

Comment:
```suggestion
Error wrapSYCLBinaries(llvm::Module &M, const SmallVector<SYCLImage> &Images,
```

?

Reply:
This is made this way for a possibility to free Images on the fly in `createFatbinDesc`.

Reply:
Hmm. Will this not cause ownership issues? We would always want Images to be freed in the same function where it was created, right?